DeepDiveLogCompiler.scala

```diff
@@ -170,29 +177,68 @@ class CompilationState( statements : DeepDiveLog.Program )  {
         ${ whereClauseStr }"""
   }
 
+  // Analyze the block visibility among statements 
+  def analyzeVisable(statements: List[Statement]) = {
+    val extractionRules   = new ListBuffer[ExtractionRule]()
+    val functionCallRules = new ListBuffer[FunctionCallRule]()
+    val inferenceRules    = new ListBuffer[InferenceRule]()
+    
+    statements foreach (_ match {
+      case s: ExtractionRule   => extractionRules += s
+      case s: FunctionCallRule => functionCallRules += s
+      case s: InferenceRule    => inferenceRules += s
+      case _                   => 
+    })
+
+    val extractionRulesGroup   = extractionRules.groupBy(_.q.head.name)
+    val functionCallRulesGroup = functionCallRules.groupBy(_.input)
+    val inferenceRulesGroup    = inferenceRules.groupBy(_.q.head.name)
+
+    extractionRulesGroup   foreach {keyVal => visable += keyVal._2(0)}
+    functionCallRulesGroup foreach {keyVal => visable += keyVal._2(0)}
+    inferenceRulesGroup    foreach {keyVal => visable += keyVal._2(0)}
+  }
```

Comment:
Sure, please see the new commit. I just keep Map[String, List[Statement]] because it's easy to groupby. Do you have better way to groupby head and generate a Map[Statement, List[Statement]]?